eva-accession-import/src/main/java/uk/ac/ebi/eva/accession/dbsnp/contig/ContigMapping.java

```diff
@@ -20,50 +20,98 @@
 import uk.ac.ebi.eva.accession.dbsnp.io.AssemblyReportReader;
 
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
 public class ContigMapping {
 
-    private Map<String, ContigSynonyms> sequenceNameToSynonyms;
+    private Map<String, ContigSynonyms> sequenceNameToSynonyms = new HashMap<>();
 
-    private Map<String, ContigSynonyms> genBankToSynonyms;
+    private Map<String, ContigSynonyms> assignedMoleculeToSynonyms = new HashMap<>();
 
-    private Map<String, ContigSynonyms> refSeqToSynonyms;
+    private boolean assignedMoleculeHasDuplicates = false;
 
-    private Map<String, ContigSynonyms> ucscToSynonyms;
+    private Map<String, ContigSynonyms> genBankToSynonyms = new HashMap<>();
+
+    private Map<String, ContigSynonyms> refSeqToSynonyms = new HashMap<>();
+
+    private Map<String, ContigSynonyms> ucscToSynonyms = new HashMap<>();
 
     private static final String CHROMOSOME_PATTERN = "(chromosome|chrom|chr|ch)(.+)";
 
     private static final Pattern PATTERN = Pattern.compile(CHROMOSOME_PATTERN, Pattern.CASE_INSENSITIVE);
 
+    private static final String NOT_AVAILABLE = "na";
+
     public ContigMapping(String assemblyReportUrl) throws Exception {
         this(new AssemblyReportReader(assemblyReportUrl));
     }
 
     public ContigMapping(AssemblyReportReader assemblyReportReader) throws Exception {
-        sequenceNameToSynonyms = new HashMap<>();
-        genBankToSynonyms = new HashMap<>();
-        refSeqToSynonyms = new HashMap<>();
-        ucscToSynonyms = new HashMap<>();
         populateMaps(assemblyReportReader);
     }
 
+    public ContigMapping(List<ContigSynonyms> contigSynonyms) {
+        contigSynonyms.forEach(this::fillContigConventionMaps);
+        if (assignedMoleculeHasDuplicates) {
+            removeAssignedMolecule();
+        }
+    }
+
+    private void removeAssignedMolecule() {
+        assignedMoleculeToSynonyms.clear();
+        sequenceNameToSynonyms.forEach((key, contigSynonym) -> contigSynonym.setAssignedMolecule(null));
+        ucscToSynonyms.forEach((key, contigSynonym) -> contigSynonym.setAssignedMolecule(null));
+        genBankToSynonyms.forEach((key, contigSynonym) -> contigSynonym.setAssignedMolecule(null));
+        refSeqToSynonyms.forEach((key, contigSynonym) -> contigSynonym.setAssignedMolecule(null));
+    }
+
     private void populateMaps(AssemblyReportReader assemblyReportReader) throws Exception {
         ContigSynonyms contigSynonyms;
         while ((contigSynonyms = assemblyReportReader.read()) != null) {
             fillContigConventionMaps(contigSynonyms);
         }
+        if (assignedMoleculeHasDuplicates) {
```

Comment:
Isn't it a bit extreme to drop **all** the assigned molecules as soon as just one of them has a duplicate?

Reply:
Can we rely on the order of the rows on the assembly_reports and guarantee that the first appearance is always the best?

Reply:
All the reports we have seen so far list the chromosomes first, but can't be 100% sure if that is always the case.